### PR TITLE
BugFix：1.ON DUPLICATE KEY UPDATE Sub-Clause paramter dropped   https:…

### LINKED ISSUE #5210: ON DUPLICATE KEY UPDATE Sub-Clause paramter dropped
DAO层使用 MyBatis，当使用 INSERT INTO VALUES 
ON DUPLICATE KEY UPDATE 
语句时，UPDATE 部分参数全都会丢失

### PR DESCRIPTION
**Pull request by neil4dong**
Fixes #5210.

Changes proposed in this pull request:
- Add assistedQueryColumn logic to ON DUPLICATE KEY UPDATE Sub-Clause.   and unit test with it.
- BugFix:  ON DUPLICATE KEY UPDATE Sub-Clause paramter dropped  
 https://github.com/apache/shardingsphere/issues/5210
-

### CODE REVIEW

**Comment by tristaZero** [comment]
@neil4dong Hi thanks for your effort in this PR. I will give it a careful look at it, please leave me some time. :)

**Comment by tristaZero** [comment]
@neil4dong Hi it is a major change, and IMO more detail needs to discuss. Do you mind exchanging our WeChat number to give it a discussion? If not, could you send your WeChat number to `panjuan@apache.org`?

File: shardingsphere-underlying/shardingsphere-rewrite/shardingsphere-rewrite-engine/src/main/java/org/apache/shardingsphere/underlying/rewrite/engine/RouteSQLRewriteEngine.java [hunk 1]
```diff
@@ -62,7 +62,10 @@ public RouteSQLRewriteResult rewrite(final SQLRewriteContext sqlRewriteContext,
         int count = 0;
         for (Collection<DataNode> each : routeResult.getOriginalDataNodes()) {
             if (isInSameDataNode(each, routeUnit)) {
-                result.addAll(((GroupedParameterBuilder) parameterBuilder).getParameters(count));
+                GroupedParameterBuilder groupedParameterBuilder = (GroupedParameterBuilder) parameterBuilder;
+
+                result.addAll(groupedParameterBuilder.getParameters(count));
+                result.addAll(groupedParameterBuilder.getOriginalOnDuplicateKeyParameters());
```
Comment (neil4dong) [inline shardingsphere-underlying/shardingsphere-rewrite/shardingsphere-rewrite-engine/src/main/java/org/apache/shardingsphere/underlying/rewrite/engine/RouteSQLRewriteEngine.java:68]: This line will lead  parameters miss match when sharding to mutiple datanodes..

File: shardingsphere-underlying/shardingsphere-rewrite/shardingsphere-rewrite-engine/src/main/java/org/apache/shardingsphere/underlying/rewrite/engine/RouteSQLRewriteEngine.java [hunk 2]
```diff
@@ -62,7 +62,10 @@ public RouteSQLRewriteResult rewrite(final SQLRewriteContext sqlRewriteContext,
         int count = 0;
```
Comment (neil4dong) [inline shardingsphere-underlying/shardingsphere-rewrite/shardingsphere-rewrite-engine/src/main/java/org/apache/shardingsphere/underlying/rewrite/engine/RouteSQLRewriteEngine.java:62]: Above this line, there is return early style code,   Should exclude when parameters is empty and onduplicateparameter is not empty.

File: shardingsphere-underlying/shardingsphere-rewrite/shardingsphere-rewrite-engine/src/main/java/org/apache/shardingsphere/underlying/rewrite/engine/RouteSQLRewriteEngine.java (hunk 1 above)
Comment (neil4dong) [reply shardingsphere-underlying/shardingsphere-rewrite/shardingsphere-rewrite-engine/src/main/java/org/apache/shardingsphere/underlying/rewrite/engine/RouteSQLRewriteEngine.java:68]: added new test case for this problem.

**Pull request closed without merging**